Manager experiment

ajCameron opened the pull request:
Sketch of the eventual form of the manager is now complete.

Additional capability needed to be added, but a review of the direction would be appreciated.

ajCameron committed:
Manager skeleton added

ajCameron committed:
Starting to sketch in methods

 - But discord input seems to be broken. Need to pause and investigate that.

ajCameron committed:
Fix to registry linting

ajCameron committed:
Manager sketched in

 - Input side of manager interactions with IOConfigs added
 - ManagerInput events created
 - Checking for manager command trigger prototype added
 - all linters and tests passing - for now
 - Manager now needs to be incorporated into the bot runner and tested

ajCameron committed:
Continuing to work in the draft of the manager

 - Probably time for a confab to make sure it's working in a vaguely sane way.

javajawa committed:
mess around and find out

javajawa committed:
Add more manager notes and concepts

ajCameron committed:
Manager in place and input working

 - ManagerInterface established
 - Manager established
 - manager incorporated into bot
 - manager trigger methodology estblished
 - ManagerInput/Output events established
 - No tests done so far - but linting is passing

ajCameron committed:
Working manager prototype is up

 - skeleton of all the manager is in place
 - discord has been extended to include a initial manager interface
 - !status now works - yields a reply containing the status
 - some tidying up to do - but the basic interface has been established.

ajCameron committed:
Preliminary tests added

 - Definitions complete
 - sketch of manager program flow complete
 - tests cover as much of the work so far as can be done without mocking in more inputs (idea - add a zmq input to assist in testing - as that can all be run internally)
 - additional capacity can be added, but review of the direction would probably be good at this point

exachixkitsune commented on dev-docs/component-dev-notes.md:
```diff
@@ -0,0 +1,30 @@
+The intent of this is that the system knows what objects have been registered so it can present to the user on an API level all the components which can be included in a bot configuration.
+However, setting this up does mean adding some definitions in a couple of different places.
+
+### If you are changing a declared API
```
Probably want to consider a "if you're changing an API, increment the verno as per this scheme"; or at least a reference here to the other document which says how that's done. An api-dev-notes.md?

ajCameron replied:
I have added a reference to api-dev-notes.md - will construct.

exachixkitsune commented on dev-docs/component-dev-notes.md:
```diff
@@ -0,0 +1,30 @@
+The intent of this is that the system knows what objects have been registered so it can present to the user on an API level all the components which can be included in a bot configuration.
+However, setting this up does mean adding some definitions in a couple of different places.
+
+### If you are changing a declared API
```
Do we want to say "<these> changes do not require an increment to the api". Probably stuff along the lines of "you are purely adding new features not modifying old ones"

ajCameron replied:
This seems clearer - updated

exachixkitsune reviewed (dismissed):
Obviously needs the conflict fixing.
Code not reviewed in great detail, but not seen any glaring issues.

ajCameron committed:
Merge branch 'mainline' into manager_experiment

ajCameron committed:
Updating project to account for sonarcloud requirements

 - Breaking down bot and manager to smaller components for easier testing
 - Adding doc strings - mostly to reduce the testing load
 - Added tools to allow for more direct testing of bots
 - Shifted where the token is stored in discord input to reduce public attributes
 - Removed rogue print statements
 - Enhanced discord testing - but more work still needed

ajCameron committed:
Merge branch 'mainline' into manager_experiment
